### PR TITLE
OCS Share API to appframework

### PR DESCRIPTION
**Pull request by rullzer**
First OCS API to be ported.

TODO:
- [X] Fix tests
- [X] Add middleware
- [X] Add more response types (I don't like those magic arrays retruned)
- [x] Test middleware
- [x] Split PR into OCSMiddleware thingy

### CODE REVIEW

**Comment by rullzer** [comment]
The thing here is that we can't use all (we can use some) of the standard AppFramework responses. Since here we want to add a message often as well (to not change the output of the API).

Should we add OCSNotFoundResponse etc?
Other idea's?

@BernhardPosselt @LukasReschke @icewind1991 @nickvergessen @MorrisJobke

**Comment by BernhardPosselt** [comment]
Depends really. You could use a trait like https://github.com/owncloud/app-tutorial/blob/master/controller/errors.php#L13 or implement an afterException hook in a middleware that responds to an OCSNotFoundException, you could add a shortcut function, custom response type etc etc :D

**Comment by rullzer** [comment]
Fair enough.
But I'm more looking for something that we could use in all cases where the OCS API is in use.

Maybe an OCSMiddleware that lives in core would not be so bad. Then it just catches OCS***Exceptions.

**Comment by BernhardPosselt** [comment]
There's definitely a benefit in adding a type for these cases since it makes it harder to make mistakes. Types and Exceptions are sometimes hard to distinguish and other languages actually only have types (e.g. Result<String, Error> where String is the success case and Error the error case)

Ultimately I like the exception approach a little bit more than the custom response because I think its more idiomatic.

**Comment by rullzer** [comment]
After thinking about it I think I prefer the exceptions as well.

I'll create a PR soonish and let you know.

**Comment by rullzer** [comment]
Rebase on https://github.com/nextcloud/server/pull/475 once in...

**Comment by rullzer** [comment]
Needs: https://github.com/nextcloud/server/pull/480

**Comment by rullzer** [comment]
Ok finally time to review. A lot is rewriting the tests. But I think this cleaned up the code alread pretty good.

CC: @schiessle @nickvergessen @MorrisJobke @LukasReschke

File: apps/files_sharing/tests/ApiTest.php [hunk 1]
```diff
@@ -291,8 +297,13 @@ function testEnfoceLinkPassword() {
 
 		$request = $this->createRequest($data);
 		$ocs = $this->createOCS($request, self::TEST_FILES_SHARING_API_USER1);
-		$result = $ocs->createShare();
-		$this->assertFalse($result->succeeded());
+		try {
+			$ocs->createShare();
+			$this->fail();
```
Comment (MorrisJobke) [inline apps/files_sharing/tests/ApiTest.php]: I like that one 👍

File: apps/files_sharing/tests/ApiTest.php [hunk 2]
```diff
@@ -477,9 +499,9 @@ function testPublicLinkUrl() {
 		$request = $this->createRequest($post);
 		$ocs = $this->createOCS($request, self::TEST_FILES_SHARING_API_USER1);
 		$result = $ocs->createShare();
-		$this->assertTrue($result->succeeded());
+		$ocs->cleanup();
 
-		$data = $result->getData();
+		$data = $result->getData()['data'];
```
Comment (MorrisJobke) [inline apps/files_sharing/tests/ApiTest.php]: If somebody is wondering: This was added in 5.4 ;9

**Comment by MorrisJobke** [comment]
I get following when opening the share sidebar and it logs me out, because it thinks the session has timed out:

bildschirmfoto 2016-07-25 um 09 40 19

**Comment by rullzer** [comment]
@MorrisJobke ah mmm I get that to. Let me check why.

**Comment by rullzer** [comment]
@MorrisJobke ok fixed!

**Comment by MorrisJobke** [comment]
* [ ] Bug: share by link, set a password and use `abc`

It should look like (master):

bildschirmfoto 2016-07-26 um 10 20 27

HTTP stauts: 400

But it looks like this:

bildschirmfoto 2016-07-26 um 10 21 06

HTTP status: 200

**Comment by rullzer** [comment]
will take a look

**Comment by rullzer** [comment]
@MorrisJobke can you double check? It works here as expected... (I get a 400)

**Comment by MorrisJobke** [comment]
> @MorrisJobke can you double check? It works here as expected... (I get a 400)

Yes, it is still happening. Should we do a screen sharing session?

**Comment by MorrisJobke** [comment]
Together with #606 this works fine 👍

**Comment by MorrisJobke** [comment]
@schiessle Please review

**Comment by MorrisJobke** [comment]
@schiessle @LukasReschke @nickvergessen Please review this one :)

File: apps/files_sharing/lib/API/Share20OCS.php [hunk 3]
```diff
@@ -175,96 +187,96 @@ protected function formatShare(\OCP\Share\IShare $share) {
...
 
-		return new \OC_OCS_Result(null, 404, $this->l->t('Wrong share ID, share doesn\'t exist'));
+		throw new OCSNotFoundException($this->l->t('Wrong share ID, share doesn\'t exist'));
 	}
 
 	/**
 	 * Delete a share
 	 *
+	 * @NoAdminRequired
+	 * @NoCSRFRequired
```
Comment (LukasReschke) [inline apps/files_sharing/lib/API/Share20OCS.php]: This is a bit problematic. Let me think about it.

Comment (LukasReschke) [inline apps/files_sharing/lib/API/Share20OCS.php]: Yeah. That will allow CSRF, we need to think about some more enhanced logic for this OCSControllers. This is a little bit of a PITA due to the fact that we can't enforce simply token auth or so (which avoids the risk of CSRF).

We probably need to add a middleware that checks if the request header `OCS-APIREQUEST` is set to `true` if authentication cookies are sent with the request. Basic authentication without any provided cookies should still work. An even better way would be to have the token stuff mentioned above, but I guess step by step we get there :-)

My personal preference would be to handle this completely transparently in the middleware and the developer should have not to do any CSRF arguments here.

Comment (rullzer) [inline apps/files_sharing/lib/API/Share20OCS.php]: Yep.
So few things that come to mind:

* We can check in the CSRF middleware part if something is an OCSController if not we don't do this magic at all
* If so do the stuff you describe

And yeah token auth will be better. (But then we also need to store the token for the browser session).

**Comment by rullzer** [comment]
Lets wait for https://github.com/nextcloud/server/pull/656 to be in

**Comment by MorrisJobke** [comment]
Rebased on master.

@LukasReschke @rullzer Is the above mentioned issue fixed with this then?

**Comment by rullzer** [comment]
@MorrisJobke it is now.
@LukasReschke any other remarks?

**Comment by rullzer** [comment]
Ah joy this breaks the intergration tests... because we did not set the header there.

**Comment by rullzer** [comment]
All fixed.

@LukasReschke @schiessle @nickvergessen final review?

**Comment by MorrisJobke** [comment]
> This branch has conflicts that must be resolved

@rullzer

**Comment by rullzer** [comment]
Conflicts resolved.
Review time!!!

File: apps/files_sharing/lib/API/Share20OCS.php [hunk 4]
```diff
@@ -471,25 +466,24 @@ private function getSharesInDir($folder) {
...
 
-		return new \OC_OCS_Result($formatted);
+		return new DataResponse(['data' => $formatted]);
 	}
 
 	/**
 	 * The getShares function.
 	 *
+	 * @NoAdminRequired
+d	 *
```
Comment (LukasReschke) [inline apps/files_sharing/lib/API/Share20OCS.php]: 😱

Comment (rullzer) [inline apps/files_sharing/lib/API/Share20OCS.php]: Fixed

**Comment by MorrisJobke** [comment]
Still works 👍

**Comment by LukasReschke** [comment]
👍

**Pull request merged**